### PR TITLE
feat(sdk): Support loading pipeline from yaml

### PR DESCRIPTION
**Pull request by chensun**
**Description of your changes:**
- Support loading pipeline from yaml to use as a component.
- Fix bug where InputSpec ignores artifacts inputs from IR.
- Fix bug where executor label and component name may not be merged correctly due to name collision.

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

**Review by zichuan-scott-xu — COMMENTED** [review]
Thanks Chen! /lgtm

**Comment by connor-mccarthy** [comment]
Reviewing this 👀

**Review by connor-mccarthy — CHANGES_REQUESTED** [review]
Thanks, @chensun. I think this there may be an idempotency issue or at least something I have a question about.

This implementation results in changing PipelineSpec when you reload the pipeline. Try appending the following to one of the pipeline files, running it, then inspecting the `one.yaml` and `two.yaml` files. They are different.

```python
if __name__ == '__main__':
    compiler.Compiler().compile(
        pipeline_func=my_pipeline,
        package_path='one.yaml')

    from kfp import components
    with open('one.yaml', 'r') as f:
        reloaded_pipeline = components.load_component_from_text(f.read())
    compiler.Compiler().compile(
        reloaded_pipeline, package_path='two.yaml')
```

I don't think this is desired, correct? Even if the underlying pipeline logic is itself correct. WDYT?

If I am correct, the read/write tests will be very helpful for validating correctness across a diversity of pipelines if you switch the `.pipelines.config.read` flag to `True`, plus make a few other changes (see this commit for the additional changes needed: https://github.com/connor-mccarthy/pipelines/commit/746d787277e396ebf27b2d484878c2c2eb4c4333). If I am not correct, we probably still want to use the read/write tests to validate correctness, but they will need more attention in order to do so.

**Comment by chensun** [comment]
> 



> Thanks, @chensun. I think this there may be an idempotency issue or at least something I have a question about.
> 
> This implementation results in changing PipelineSpec when you reload the pipeline. Try appending the following to one of the pipeline files, running it, then inspecting the `one.yaml` and `two.yaml` files. They are different.
> 
> ```python
> if __name__ == '__main__':
>     compiler.Compiler().compile(
>         pipeline_func=my_pipeline,
>         package_path='one.yaml')
> 
>     from kfp import components
>     with open('one.yaml', 'r') as f:
>         reloaded_pipeline = components.load_component_from_text(f.read())
>     compiler.Compiler().compile(
>         reloaded_pipeline, package_path='two.yaml')
> ```
> 
> I don't think this is desired, correct? Even if the underlying pipeline logic is itself correct. WDYT?
> 
> If I am correct, the read/write tests will be very helpful for validating correctness across a diversity of pipelines if you switch the `.pipelines.config.read` flag to `True`, plus make a few other changes (see this commit for the additional changes needed: [connor-mccarthy@746d787](https://github.com/connor-mccarthy/pipelines/commit/746d787277e396ebf27b2d484878c2c2eb4c4333)). If I am not correct, we probably still want to use the read/write tests to validate correctness, but they will need more attention in order to do so.

That's a good point. Although the changed PipelineSpec is still correct, agree that it is ideal to maintain consistency across read and write. Made the change.

**Comment by chensun** [comment]
All test passed. Merging this to unblock my next PR. 
@connor-mccarthy please feel free to add additional comments, I'll address them in a follow up PR if needed. Thanks!

File: sdk/python/kfp/components/yaml_component.py [hunk 1]
```diff
@@ -15,20 +15,43 @@
...
+import yaml
 
 
 class YamlComponent(components.BaseComponent):
     """A component loaded from a YAML file.
 
     **Note:** ``YamlComponent`` is not intended to be used to construct components directly. Use ``kfp.components.load_component_from_*()`` instead.
 
-    Args:
+    Artribute:
```
Comment (connor-mccarthy) [inline sdk/python/kfp/components/yaml_component.py:31]:
```suggestion
    Attributes:
```

File: sdk/python/kfp/components/structures.py [hunk 2]
```diff
@@ -511,18 +523,16 @@ def transform_outputs(self) -> None:
     def validate_placeholders(self):
         """Validates that input/output placeholders refer to an existing
         input/output."""
-        implementation = self.implementation
-        if getattr(implementation, 'container', None) is None:
+        if self.implementation.container is None:
```
Comment (connor-mccarthy) [inline sdk/python/kfp/components/structures.py:526]: 👍🏻

File: sdk/python/kfp/compiler/pipeline_spec_builder.py [hunk 3]
```diff
@@ -1470,12 +1488,11 @@ def _rename_component_refs(
         sub_pipeline_spec.root)
 
 
-def create_pipeline_spec_and_deployment_config(
+def create_pipeline_spec(
```
Comment (connor-mccarthy) [inline sdk/python/kfp/compiler/pipeline_spec_builder.py:1491]: I like this change

**Review by connor-mccarthy — COMMENTED** [review]
This looks great, @chensun. Thank you for updating the tests!

/lgtm

**Pull request merged**